lib/url.js

```diff
@@ -685,40 +497,609 @@ Url.prototype.resolveObject = function(relative) {
+    }
+  }
+  if (cur < i + 1) ret += str.slice(cur, i);
+  return ret;
+};
+
+Url.prototype._parsePath = function(str, start, end) {
+  var pathStart = start;
+  var pathEnd = end;
+  var escape = false;
+  var autoEscapeCharacters = _autoEscapeCharacters;
+  var prePath = this._port === -2 ? '/:' : '';
+
+  for (var i = start; i <= end; ++i) {
+    var ch = str.charCodeAt(i);
+
+    if (ch === 0x23 /*'#'*/) {
+      this._parseHash(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (ch === 0x3F /*'?'*/) {
+      this._parseQuery(str, i, end);
+      pathEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1) {
+      escape = true;
+    }
+  }
+
+  if (pathStart > pathEnd) {
+    this.pathname = prePath === '' ? '/' : prePath;
+    return;
+  }
+
+  var path;
+  if (escape) {
+    path = this._getComponentEscaped(str, pathStart, pathEnd, false);
+  } else {
+    path = str.slice(pathStart, pathEnd + 1);
+  }
+
+  this.pathname = prePath === '' ?
+      (this._prependSlash ? '/' + path : path) : prePath + path;
+};
+
+Url.prototype._parseQuery = function(str, start, end) {
+  var queryStart = start;
+  var queryEnd = end;
+  var escape = false;
+  var autoEscapeCharacters = _autoEscapeCharacters;
+
+  for (var i = start; i <= end; ++i) {
+    var ch = str.charCodeAt(i);
+
+    if (ch === 0x23 /*'#'*/) {
+      this._parseHash(str, i, end);
+      queryEnd = i - 1;
+      break;
+    } else if (!escape && autoEscapeCharacters[ch] === 1)
+      escape = true;
+  }
+
+  if (queryStart > queryEnd) {
+    this.search = '';
+    return;
+  }
+
+  var query = escape ?
+      this._getComponentEscaped(str, queryStart, queryEnd, true) :
+      str.slice(queryStart, queryEnd + 1);
+
+  this.search = query;
+};
+
+Url.prototype._parseHash = function(str, start, end) {
+  if (start > end) {
+    this.hash = '';
+    return;
+  }
+  this.hash = this._getComponentEscaped(str, start, end, true);
+};
+
+Object.defineProperty(Url.prototype, 'port', {
+  get: function() {
+    if (this._hasValidPort())
+      return this._port;
+    return null;
+  },
+  set: function(v) {
+    if (v == null) {
+      this._port = -1;
+    } else {
+      v = parseInt(v, 10);
+      if (!isFinite(v))
+        v = 0;
+
+      v = Math.max(0, v) % 0x10000;
+      this._port = '' + v;
+    }
+  },
+  enumerable: true,
+  configurable: true
+});
+
+Object.defineProperty(Url.prototype, 'query', {
+  get: function() {
+    var query = this._query;
+    if (query != null)
+      return query;
+
+    var search = this.search;
+
+    if (search) {
+      if (search.charCodeAt(0) === 0x3F /*'?'*/)
+        search = search.slice(1);
+      if (search !== '') {
+        this._query = search;
+        return search;
+      }
+    }
+    return search;
+  },
+  set: function(v) {
+    if (typeof v === 'string') {
+      if (v !== '') this.search = '?' + v;
+      this._query = v;
+    } else if (v !== null && typeof v === 'object') {
```

Comment:
Did we seriously allow setting to an object before?? If not, we should not add that behavior.